flash/image/classification/integrations/baal/data.py

```diff
@@ -0,0 +1,161 @@
+from flash.core.data.data_pipeline import DataPipeline
+from flash.core.utilities.imports import _BAAL_AVAILABLE, requires
+
+if _BAAL_AVAILABLE:
+    from baal.active.dataset import ActiveLearningDataset
+    from baal.active.heuristics import AbstractHeuristic, BALD
+else:
+
+    class AbstractHeuristic:
+        pass
+
+    class BALD(AbstractHeuristic):
+        def __init__(self, reduction: Callable):
+            super().__init__()
+
+
+def dataset_to_non_labelled_tensor(dataset: BaseAutoDataset) -> torch.tensor:
+    return torch.zeros(len(dataset))
+
+
+def filter_unlabelled_data(dataset: BaseAutoDataset) -> Dataset:
+    return dataset
+
+
+def train_val_split(dataset: Dataset, val_size=0.1):
+    L = len(dataset)
+    train_size = int(L * (1 - val_size))
+    val_size = L - train_size
+    return random_split(dataset, [train_size, val_size], generator=torch.Generator().manual_seed(42))
+
+
+class ActiveLearningDataModule(LightningDataModule):
+    @requires("baal")
+    def __init__(
+        self,
+        labelled: Optional[DataModule] = None,
+        heuristic: "AbstractHeuristic" = BALD(reduction=np.mean),
+        map_dataset_to_labelled: Optional[Callable] = dataset_to_non_labelled_tensor,
+        filter_unlabelled_data: Optional[Callable] = filter_unlabelled_data,
+        num_label_randomly: int = 5,
+        val_split: Optional[float] = None,
+    ):
+        """The `ActiveLearningDataModule` handles data manipulation for ActiveLearning.
+
+        Args:
+            labelled: DataModule containing labelled train data for research use-case.
+                The labelled data would be masked.
+            heuristic: Sorting algorithm used to rank samples on how likely they can help with model performance.
+            map_dataset_to_labelled: Function used to emulate masking on labelled dataset.
+            filter_unlabelled_data: Function used to filter the unlabelled data while computing uncertainties.
+            num_label_randomly: Number of samples to randomly labelled from the uncertaintie scores
+            val_split: Float to split train dataset into train and validation set.
+        """
+        self.labelled = labelled
+        self.heuristic = heuristic
+        self.map_dataset_to_labelled = map_dataset_to_labelled
+        self.filter_unlabelled_data = filter_unlabelled_data
+        self.num_label_randomly = num_label_randomly
+        self.val_split = val_split
+        self._dataset: Optional[ActiveLearningDataset] = None
+
+        if not self.labelled:
+            raise MisconfigurationException("The labelled `datamodule` should be provided.")
+
+        if not self.labelled.num_classes:
+            raise MisconfigurationException("The labelled dataset should be labelled")
+
+        if self.labelled and (
+            self.labelled._val_ds is not None
+            or self.labelled._test_ds is not None
+            or self.labelled._predict_ds is not None
+        ):
```

Comment:
I agree with the API here, i.e. force the user to provide only the train dataset. 

However, for the "research case" it would be nice to be able to provide the test dataset as well. This would allow us to easily test within the active learning loop and create the common "training budget vs test accuracy" graphs that are often reported in papers (example attached below).

What do you think?


![active-learning-fig2](https://user-images.githubusercontent.com/61748653/134482576-51ae1eec-f44d-4dee-8c3a-94440a9d2baa.png)
)

Reply:
Fully agree with @pietrolesci here to prove out the active learning case! maybe this can be a separate PR?

Reply:
Done !